keps/sig-scheduling/20190221-even-pods-spreading.md

```diff
@@ -0,0 +1,386 @@
+next the app needs to perform a rolling upgrade, then a third replacement pod is
+created, but it failed to be placed due to lack of resource. In this case,
+
+- if CA is enabled, a new machine will be provisioned to hold the new pod
+  (although old replicas will be deleted afterwards) (see
+  [#40358](https://github.com/kubernetes/kubernetes/issues/40358))
+- if CA is not enabled, it's a deadlock since the replacement pod can't be
+  placed. The only workaround at this moment is to update app strategyType from
+  "RollingUpdate" to "Recreate".
+
+Neither of them is an ideal solution. A promising solution is to give user an
+option to trigger "toleration" mode when the cluster is out of resource. Then in
+aforementioned example, a third pod is "tolerated" to be put onto node1 (or
+node2). But keep it in mind, this behavior is only triggered upon resource
+shortage. For a 3-nodes cluster, the third pod will still be placed onto node3
+(if node3 is capable).
+
+### Risks and Mitigations
+
+The feature requires additional processing for pods that use it and it is ok to
+have some performance overhead. But we will make sure our implementation will
+not have any performance penalty for pods that do not use this feature.
+
+## Design Details
+
+### API
+
+A new structure called `EvenSpreading` is introduced which acts as a standalone
+spec and applies to `pod.spec`. It's only effective when it's not nil.
+
+```go
+type PodSpec struct {
+    EvenSpreading *EvenSpreading
+    ......
+}
+```
+
+#### Option 1
+
+Inside `EvenSpreading`, we need hard affinityTerms (similar with
+`PodAffinityTerm`) and soft affinityTerms (similar with
+`WeightedPodAffinityTerm`). This describes when we perform even distribution,
+which pods are considered as a group.
+
+```go
+type EvenSpreading struct {
+    // MaxSkew describes the degree of imbalance of pods spreading.
+    // It's the max difference between the number of matching pods in any two
+    // topology domains of a given topology type.
+    // Default value is 1 and 0 is not allowed.
+    MaxSkew int32
+    // TopologyKey defines where pods are placed evenly
+    TopologyKey string
+    // Similar with the same field in PodAffinity/PodAntiAffinity
+    // +optional
+    RequiredDuringSchedulingIgnoredDuringExecution []PodAffinityTerm
+    // Similar with the same field in PodAffinity/PodAntiAffinity
+    // +optional
+    PreferredDuringSchedulingIgnoredDuringExecution []WeightedPodAffinityTerm
+}
+```
+
+#### Option 2
+
+Another option is to flatten "required" and "perferred" podAffinityTerms, and
+eliminate embedded "TopologyKey":
+
+```go
+type PodSelectorTerm struct {
+	  // A label query over a set of resources, in this case pods.
+	  // +optional
+	  LabelSelector *metav1.LabelSelector
+	  // namespaces specifies which namespaces the labelSelector applies to (matches against);
+	  // null or empty list means "this pod's namespace"
+	  // +optional
+	  Namespaces []string
```

Comment:
Why is it necessary to have label selectors in different namespaces? The user stories don't demonstrate that it is necessary. If it is not necessary, I'd prefer to represent this whole thing with a single label selector rather than permitting many.

Reply:
I basically agree.

The reason that podAffinity has `[]podAffinityTerms` is because inside individual "podAffinityTerm", it has its own tuple of {labelSelector, namespaces and topologyKey}. And here we've almost agreed to eliminate embedded "TopologyKey", and hence should also be reasonable to merge all labelSelector into one (merging semantics works here).

In terms of "Namespaces", it can be either (1) flatten out to be as the same level of the unique labelSelector, or (2) removed as it can be defaulted to "the pod's namespace". I'm open to either option.

/cc @bsalamat